### PR TITLE
Make expression manipulation consistent and easier to use: `combine/split filter` `conjunction`, etc

### PR DESCRIPTION
**Pull request by alamb**
~DRAFT as it builds on https://github.com/apache/arrow-datafusion/pull/3809~

# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/3808

 # Rationale for this change

The APIs for manipulating expressions were all over the map (sometimes return Vec, sometimes taking them as mut, owned, etc) as well as being inconsistently named and inconsistently tested. 

In fact I couldn't find `uncombine_filter` (which is similar, but not the same as `split_conjunction`)

# What changes are included in this PR?
Change the APIs to be consistently named and take reasonable arguments

1. Change `split_conjunction` to return a `Vec` thus simplifying the API
2. Rename split_filter to `split_conjunction_owned` to make it clear how it is related to `split_conjunction`
3. Expand test coverage
2. Rename `combine_filter` to `conjunction`
2. Rename `combine_filter_disjunction` to `disjunction`
4. Change APIs for `conjunction` and `disjunction` to avoid clones

I will highlight the changes inline. 

# Are there any user-facing changes?
If anyone uses these APIs they will need to change them, but hopfully

### CODE REVIEW

File: benchmarks/src/bin/tpch.rs [hunk 1]
```diff
@@ -766,7 +766,8 @@ mod tests {
             if !actual.is_empty() {
                 actual += "\n";
             }
-            actual += &format!("{}", plan.display_indent());
+            use std::fmt::Write as _;
```
Comment (alamb) [inline benchmarks/src/bin/tpch.rs:769]: clippy told me to do this

File: datafusion/optimizer/src/filter_push_down.rs [hunk 2]
```diff
@@ -341,8 +341,7 @@ fn optimize(plan: &LogicalPlan, mut state: State) -> Result<LogicalPlan> {
         }
         LogicalPlan::Analyze { .. } => push_down(&state, plan),
         LogicalPlan::Filter(Filter { input, predicate }) => {
-            let mut predicates = vec![];
-            utils::split_conjunction(predicate, &mut predicates);
+            let predicates = utils::split_conjunction(predicate);
```
Comment (alamb) [inline datafusion/optimizer/src/filter_push_down.rs:344]: here is an example where the `split_conjuntion` API is easier to use now

File: datafusion/optimizer/src/decorrelate_where_in.rs [hunk 3]
```diff
@@ -175,7 +173,7 @@ fn optimize_where_in(
     // build subquery side of join - the thing the subquery was querying
     let subqry_alias = format!("__sq_{}", optimizer_config.next_id());
     let mut subqry_plan = LogicalPlanBuilder::from((*subqry_input).clone());
-    if let Some(expr) = combine_filters(&other_subqry_exprs) {
+    if let Some(expr) = conjunction(other_subqry_exprs) {
```
Comment (alamb) [inline datafusion/optimizer/src/decorrelate_where_in.rs:176]: here is an example of less copying (can use `other_subqry_exprs`directly)

File: datafusion/optimizer/src/utils.rs [hunk 4]
```diff
@@ -51,62 +51,72 @@ pub fn optimize_children(
...
-    }
-    let combined_filter = filters
-        .iter()
-        .skip(1)
-        .fold(filters[0].clone(), |acc, filter| and(acc, filter.clone()));
-    Some(combined_filter)
+/// Splits an owned conjunctive [`Expr`] such as `A AND B AND C` => `[A, B, C]`
+///
+/// See [`split_conjunction`] for more details.
+pub fn split_conjunction_owned(expr: Expr) -> Vec<Expr> {
```
Comment (alamb) [inline datafusion/optimizer/src/utils.rs:83]: this function used to be called `uncombine_filters` which was not at all consistent with the `split_conjunction` that took a reference 🤯

File: datafusion/optimizer/src/utils.rs [hunk 5]
```diff
@@ -51,62 +51,106 @@ pub fn optimize_children(
...
             left,
         } => {
-            split_conjunction(left, predicates);
-            split_conjunction(right, predicates);
+            let exprs = split_conjunction_impl(left, exprs);
+            split_conjunction_impl(right, exprs)
         }
-        Expr::Alias(expr, _) => {
-            split_conjunction(expr, predicates);
+        Expr::Alias(expr, _) => split_conjunction_impl(expr, exprs),
```
Comment (jackwener) [inline datafusion/optimizer/src/utils.rs:71]: 👍

File: datafusion/optimizer/src/utils.rs [hunk 6]
```diff
@@ -424,31 +468,74 @@ mod tests {
...
-    use datafusion_expr::{binary_expr, col, lit, utils::expr_to_columns};
+    use datafusion_expr::{col, lit, utils::expr_to_columns};
     use std::collections::HashSet;
     use std::ops::Add;
 
     #[test]
-    fn combine_zero_filters() {
-        let result = combine_filters(&[]);
-        assert_eq!(result, None);
+    fn test_split_conjunction() {
```
Comment (jackwener) [inline datafusion/optimizer/src/utils.rs:476]: IMO, I think we can add test for `conjunction()`. At the same time, we can check the tree structure of this expression by using `match`.

```rs
expr is (A B) C;

/// using `match` to check.
match expr {
   And(
      And(
         B,
         C
      ),
      C
   )
}
```

[A, B, C ,D , E] -> (((A B) C) (D E)) is different from ((((A B) C) D) E).

we can see the result of `conjunction()` in the UT

**Review by jackwener — APPROVED** [review]
Great job👍

File: datafusion/optimizer/src/utils.rs (hunk 6 above)
Comment (alamb) [reply datafusion/optimizer/src/utils.rs:472]: This was a great idea @jackwener  -- thank you. I added this test  09f3cf4e4

As part of writing tests, I also found that the API for `disjunction` was slightly different than `conjunction` so I made them the same as well.

**Pull request merged**